### PR TITLE
Enable long paths on Tasks

### PR DESCRIPTION
**Pull request by ccastanedaucf**
Slowly working on this whenever there's time. So far this covers 2/3 tasks.

### CODE REVIEW

File: src/Shared/NativeMethodsShared.cs [hunk 1]
```diff
@@ -1241,6 +1241,48 @@ internal static int GetParentProcessId(int processId)
...
+        }
+
+        private static string GetModuleFileName(IntPtr handle)
+        {
+            bool isBufferTooSmall = false, success = false, hasNoMaxPath = OSMaxPathLimit == MaxPathLimits.None;
+            int initialSize = (int)MaxPathLimits.LegacyWindows;
+            var buffer = new StringBuilder();
+
+            // Try increased buffer sizes if on longpath-enabled Windows
+            for (int bufferSize = initialSize; hasNoMaxPath && isBufferTooSmall; bufferSize *= 2)
```
Comment (ccastanedaucf) [inline src/Shared/NativeMethodsShared.cs]: Sadly I'm not sure if there's a better way to do this with the way the native API is designed. There's no way to get the required buffer size from the function.
https://msdn.microsoft.com/en-us/library/windows/desktop/ms683197(v=vs.85).aspx

File: src/Shared/NativeMethodsShared.cs [hunk 2]
```diff
@@ -1241,6 +1241,48 @@ internal static int GetParentProcessId(int processId)
             return myChildren;
         }
 
+        internal static string GetModuleFileName(string fileName)
```
Comment (ccastanedaucf) [inline src/Shared/NativeMethodsShared.cs]: Noticed this execution path isn't covered at all by tests in ```StripTypeLibNumberFromPath```, what would be a good way to test this?

File: src/Tasks/FileState.cs [hunk 3]
```diff
@@ -115,9 +115,8 @@ public FileDirInfo(string filename)
                             //
                             // Also, when not under debugger (!) it will give error == 3 for path too long. Make that consistently throw instead.
                             if ((error == 2 /* ERROR_FILE_NOT_FOUND */|| error == 3 /* ERROR_PATH_NOT_FOUND */)
-                                && _filename.Length <= NativeMethodsShared.MAX_PATH)
+                                && _filename.Length <= (int)NativeMethodsShared.OSMaxPathLimit)
```
Comment (cdmihai) [inline src/Tasks/FileState.cs]: Why the cast?

Comment (ccastanedaucf) [reply src/Tasks/FileState.cs]: Enum value, maybe we should just split up OSMaxPathLimit into a bool and int value if the only real case to watch for is Windows + 260.

File: src/Shared/NativeMethodsShared.cs [hunk 4]
```diff
@@ -48,7 +48,7 @@ internal static class NativeMethodsShared
         /// <summary>
         /// Default buffer size to use when dealing with the Windows API.
         /// </summary>
-        internal const int MAX_PATH = 260;
+        private const int MAX_PATH = 260;
```
[the remark this answers is not in the document]
Comment (ccastanedaucf) [reply src/Shared/NativeMethodsShared.cs]: Actually was my original name for it https://github.com/Microsoft/msbuild/pull/3406#discussion_r196579408

File: src/Shared/NativeMethodsShared.cs [hunk 5]
```diff
@@ -467,44 +460,55 @@ public SystemInformationData()
...
+        internal static int MaxPath
         {
             get
             {
 #if EXPERIMENTAL_LONGPATHS_ENABLED
-                if (osMaxPathLimit == MaxPathLimits.Unknown)
+                if (!isMaxPathSet)
                 {
-                    SetOSMaxPathLimit();
+                    SetMaxPath();
```
Comment (cdmihai) [inline src/Shared/NativeMethodsShared.cs]: put the ifdef inside SetMaxPath, to have all initialization logic in one place.

File: src/Tasks/FileState.cs [hunk 6]
```diff
@@ -101,6 +101,9 @@ public FileDirInfo(string filename)
                 {
                     if (NativeMethodsShared.IsWindows)
                     {
+                        // GetFileAttributesEx() will fail with long paths if not normalized
+                        _filename = NativeMethodsShared.HasMaxPath ? _filename : FileUtilities.NormalizePath(_filename);
```
Comment (cdmihai) [inline src/Tasks/FileState.cs]: Judging by the comment, shouldn't the path get normalized only when there are max path limits?

**Comment by ccastanedaucf** [comment]
So this passes locally and I'm not seeing how a change in ```Tasks``` would have any effect on ```ProjectGraph``` cycle detection.

```
Error message
System.Collections.Generic.KeyNotFoundException : The given key was not present in the dictionary.
Stack trace
   at System.Collections.Concurrent.ConcurrentDictionary`2.get_Item(TKey key)
   at Microsoft.Build.Graph.ProjectGraph.DetectCycles(ProjectGraphNode node, Dictionary`2 nodeState, ProjectCollection projectCollection, PropertyDictionary`1 globalProperties) in /_/src/Build/Graph/ProjectGraph.cs:line 526
   at Microsoft.Build.Graph.ProjectGraph.DetectCycles(ProjectGraphNode node, Dictionary`2 nodeState, ProjectCollection projectCollection, PropertyDictionary`1 globalProperties) in /_/src/Build/Graph/ProjectGraph.cs:line 553
   at Microsoft.Build.Graph.ProjectGraph.DetectCycles(ProjectGraphNode node, Dictionary`2 nodeState, ProjectCollection projectCollection, PropertyDictionary`1 globalProperties) in /_/src/Build/Graph/ProjectGraph.cs:line 553
   at Microsoft.Build.Graph.ProjectGraph..ctor(IEnumerable`1 entryPoints, ProjectCollection projectCollection, ProjectInstanceFactoryFunc projectInstanceFactory) in /_/src/Build/Graph/ProjectGraph.cs:line 226
   at Microsoft.Build.Graph.ProjectGraph..ctor(String entryProjectFile) in /_/src/Build/Graph/ProjectGraph.cs:line 51
   at Microsoft.Build.Graph.UnitTests.ProjectGraphTests.ConstructWithConvergingProperties() in D:\a\1\s\src\Build.UnitTests\Graph\ProjectGraph_Tests.cs:line 279
```

**Comment by rainersigwald** [comment]
I agree, so I reran it and it passed, which . . . ok, I guess?

**Comment by ccastanedaucf** [comment]
Just talked to @yash256  and @cdmihai it seems to be a flaky test that needs fixing.

**Comment by rainersigwald** [comment]
Closing/reopening to retest after that other PR went in. If it passes I'm good to go on this.

**Pull request merged**